stock_available_mrp/models/product_product.py

```diff
@@ -16,40 +16,46 @@ class ProductProduct(models.Model):
         string='Bill of Materials'
     )
 
-    @api.depends('virtual_available')
+    @api.depends('virtual_available', 'bom_id', 'bom_id.product_qty')
     def _compute_available_quantities(self):
         super(ProductProduct, self)._compute_available_quantities()
 
     @api.multi
+    def _get_bom_id_domain(self):
+        """
+        Real multi domain
+        :return:
+        """
+        return [
+            '|',
+            ('product_id', 'in', self.ids),
+            '&',
+            ('product_id', '=', False),
+            ('product_tmpl_id', 'in', self.mapped('product_tmpl_id.id'))
+        ]
+
+    @api.multi
+    @api.depends('product_tmpl_id')
     def _compute_bom_id(self):
-        domain = [('product_id', 'in', self.ids)]
-        product_tmpl_ids = []
-        bom_product_ids = self.env['mrp.bom'].search(domain)
-        # find bom linked to a product
-        bom_by_product_id = {
-            b.product_id.id: b for b in bom_product_ids}
-        product_id_found = bom_by_product_id.keys()
-        for product in self:
-            if product.id not in product_id_found:
-                product_tmpl_ids.append(product.product_tmpl_id.id)
-        domain = [('product_id', '=', False),
-                  ('product_tmpl_id', 'in', product_tmpl_ids)]
-        # find boms linked to the product template
-        bom_product_template = self.env['mrp.bom'].search(domain)
-        bom_by_product_tmpl_id = {
-            b.product_tmpl_id.id: b for b in bom_product_template}
+        bom_obj = self.env['mrp.bom']
+        boms = bom_obj.search(
+            self._get_bom_id_domain(),
+            order='sequence, product_id',
+        )
         for product in self:
-            product.bom_id = bom_by_product_id.get(
-                product.id,
-                bom_by_product_tmpl_id.get(product.product_tmpl_id.id)
+            product_boms = boms.filtered(
+                lambda b: b.product_id == product or
+                (not b.product_id and
+                 b.product_tmpl_id == product.product_tmpl_id)
             )
+            if product_boms:
+                product.bom_id = product_boms[0]
```

Comment:
use `first()` from `from odoo.fields import first`

Reply:
Ok, never saw this helper before... They could have put it in BaseModel...